Unify Design Preview: Display both Colors & Fonts with Navigator

arthur791004 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/78419

## Proposed Changes

* Move `navigator-buttons`, `navigator-header`, and `navigator-item-group` into `@automattic/onboarding` package
* Implement `NavigatorScreens`, `useNavigatorButtons`, and `useNavigatorScreens` to display navigator buttons, and navigate between screens.
* Hide the back button on the top-left corner (next to the WP logo) when you're in either Colors or Fonts screen

### Screenshots

**Style Variations**

| Desktop | Mobile |
| - | - |
| ![image](https://github.com/Automattic/wp-calypso/assets/13596067/6a8fb6d7-87af-4105-8705-d0bf1f440a4a) | ![image](https://github.com/Automattic/wp-calypso/assets/13596067/6b330e95-ff03-44c0-9e49-7010042da87c) |

**Colors**
| Desktop | Mobile |
| - | - |
| ![image](https://github.com/Automattic/wp-calypso/assets/13596067/75ad638c-0fcf-4788-9111-018d277d5929) | ![image](https://github.com/Automattic/wp-calypso/assets/13596067/f54350af-edad-4a5c-8607-84d4eda122e4) |

**Fonts**
| Desktop | Mobile |
| - | - |
| ![image](https://github.com/Automattic/wp-calypso/assets/13596067/d5a5ec5e-6e18-4020-b988-feaca4bff8ed) | ![image](https://github.com/Automattic/wp-calypso/assets/13596067/10b60266-b1a0-47c8-a77b-bac4370409cc) |

**Both**

| Main Screen | Colors Screen | Fonts Screen |
| - | - | - |
| ![image](https://github.com/Automattic/wp-calypso/assets/13596067/7cc71c0b-69d7-4e8d-8286-08fa73d28e09) | ![image](https://github.com/Automattic/wp-calypso/assets/13596067/70e91d50-10b2-4cc4-98b0-393530e9a60d) | ![image](https://github.com/Automattic/wp-calypso/assets/13596067/fa692d4c-b1e7-4698-9ba4-f3019672965f) |

The mobile layout for both Colors & Fonts will be in the follow-up PR

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to URLs below
  * `/setup?siteSlug=<your_site>`: Fonts only
  * `/setup?siteSlug=<your_site>&flags=signup/design-picker-preview-colors`: Colors & Fonts
* Continue until you land on the Design Picker
* Preview the design with style variation, everything should be good as before
* Preview the design without the style variation
* On the preview screen,
  * You're able to see both Colors & Fonts
  * Select Colors to navigate to Colors screen
  * Select Fonts to navigate to Fonts screen
  * If you're in either Colors & Fonts screen, the back button on the top-left corner (next to the WP logo) will be hidden

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

arthur791004 committed:
Move Navigator components to @automattic/onboarding

arthur791004 committed:
Implement NavigatorScreens

arthur791004 committed:
Apply NavigatorScreens to Sidebar

arthur791004 committed:
Adjust styles

arthur791004 committed:
Hide the back button when navigating to either Colors or Fonts screen

arthur791004 committed:
Refactor to avoid screen unmounting when selected variation is changed

arthur791004 commented on packages/onboarding/src/navigator/navigator-header/style.scss:
```diff
@@ -0,0 +1,43 @@
+@import "@wordpress/base-styles/breakpoints";
```
The file is copied from the pattern assembler.

arthur791004 commented on packages/onboarding/src/navigator/navigator-header/index.tsx:
```diff
@@ -0,0 +1,40 @@
+import { Button, Gridicon } from '@automattic/components';
```
The file is copied from the pattern assembler.

arthur791004 commented on packages/onboarding/src/navigator/navigator-buttons/style.scss:
```diff
@@ -0,0 +1,62 @@
+@import "@automattic/typography/styles/fonts";
```
The file is copied from the pattern assembler.

arthur791004 commented on packages/onboarding/src/navigator/navigator-item-group/index.tsx:
```diff
@@ -6,16 +6,18 @@ import './style.scss';
 type Props = {
 	children: React.ReactNode;
-	title: string;
+	title?: string;
```
The file is copied from the pattern assembler. Make the title to be optional since we might only have a single group and it doesn't require a title.

arthur791004 commented on packages/onboarding/src/navigator/navigator-item-group/style.scss:
```diff
@@ -0,0 +1,30 @@
+@import "@automattic/typography/styles/fonts";
```
The file is copied from the pattern assembler.

arthur791004 commented on packages/onboarding/src/navigator/navigator-item-group/style.scss:
```diff
@@ -0,0 +1,30 @@
+		}
+	}
+
+	&:nth-child(1 of &):nth-last-child(1 of &) {
```
The `:only-child` doesn't select the class name, so replace it with the current way

arthur791004 commented on packages/onboarding/src/navigator/navigator-screens/hooks/use-navigator-buttons.tsx:
```diff
@@ -0,0 +1,32 @@
+	}
+
+	if ( screens.length === 1 ) {
+		return screens[ 0 ].content;
```
If there is only 1 screen, we don't need to the navigator buttons. So, return the content of the screen directly. Or should leave the parent component to determine this logic 🤔

arthur791004 commented on packages/global-styles/src/components/font-pairing-variations/preview.tsx:
```diff
@@ -2,20 +2,26 @@ import {
-	fontSize: '13.5vw', // 18px for min-width wide breakpoint and 15px for max-width wide
-	textAlign: 'center',
+const DEFAULT_LARGE_FONT_STYLES: React.CSSProperties = {
+	fontSize: '13vw', // 18px for min-width wide breakpoint and 15px for max-width wide
```
13.5vw looks a bit larger on the small screen, so reduce it to 13vw

taipeicoder commented:
Some things I noticed that we can follow up later:

1. This list of error message every time I open the theme preview.
![Screenshot 2023-07-03 at 5 28 46 PM](https://github.com/Automattic/wp-calypso/assets/797888/aa558e7a-b0ae-4f3f-beee-6ebbddfdfaa2)

2. The colors and fonts are not checked, after selection.
![Screenshot 2023-07-03 at 5 31 49 PM](https://github.com/Automattic/wp-calypso/assets/797888/e34f01ff-ad74-488f-9193-7fbf94034b73)

3. Inconsistent badges and labels

| Style variations | Colors and Fonts |
| --- | --- |
| ![Screenshot 2023-07-03 at 5 33 36 PM](https://github.com/Automattic/wp-calypso/assets/797888/fe474ea4-6df3-4f8a-bf43-fe2cbd35f5e5) | ![Screenshot 2023-07-03 at 5 33 46 PM](https://github.com/Automattic/wp-calypso/assets/797888/ff13c1dd-2c57-40f7-85c4-17115f4964c8) |

I also found that style variations is broken for some things like Pixl 🥲, but is unrelated to the PR.

taipeicoder reviewed (approved):
Nice 🎉

miksansegundo commented on packages/onboarding/src/navigator/navigator-buttons/style.scss:
```diff
@@ -0,0 +1,62 @@
+@import "@automattic/typography/styles/fonts";
+
+.components-navigator-button.navigator-button {
```
The category list in the assembler is broken, it will need the class `components-navigator-button` on [each button](https://github.com/Automattic/wp-calypso/pull/78774/files#diff-36df36d05cc8bb6d818e403469e8eb9a046feadb057e9c3d20a750235cf61fb0R107).

<img width="294" alt="Screenshot 2566-07-03 at 17 30 34" src="https://github.com/Automattic/wp-calypso/assets/1881481/9b4e0088-2fc2-4114-ab05-a9c328d82f87">

While adding the new class to `screen-category-list__category-button navigator-button` will fix it, do you have another idea to reduce the number of classnames?

arthur791004 replied:
I remembered I fixed this issue... Anyway, I'll fix it later! Nice catch!

arthur791004 replied:
Addressed it via [1446170](https://github.com/Automattic/wp-calypso/pull/78774/commits/1446170dadf0c3a4e6e5aac0400b942b2d68a71d). We can think about how to reduce the number of the classnames in the follow-up PR

miksansegundo commented:
> I also found that style variations is broken for some things like Pixl 🥲, but is unrelated to the PR.

I also noticed that the style variations are not working on most of the themes I tested.

Another issue that is unrelated to this PR is that the back button icon flashed on every render. It looks like a hint telling users to go back 😅

https://github.com/Automattic/wp-calypso/assets/1881481/aa243644-4904-483d-80f5-8817d0525c6e

It feels like the icon loads late. Maybe we can inline the SVG code rather than using a link?

https://github.com/Automattic/wp-calypso/assets/1881481/140ddf99-3ae0-47b8-9710-a53a141f8b59

arthur791004 commented:
> Another issue that is unrelated to this PR is that the back button icon flashed on every render. It looks like a hint telling users to go back 😅

It's not a bug since it's expected on the dev env. The reason is we don't enable the cache for file assets, so the assets will be requested every time. See https://github.com/Automattic/wp-calypso/pull/74279

arthur791004 committed:
Refactor via use-screens hook

arthur791004 committed:
Refactor and fix navigator button not checked

arthur791004 committed:
Fix category button in the Assembler

arthur791004 commented:
> 1. This list of error message every time I open the theme preview.

Addressed via https://github.com/Automattic/wp-calypso/pull/78925

> 2. The colors and fonts are not checked, after selection.

Added the `checked` property to the config to fix this issue 🙂

> 3. I also found that style variations is broken for some things like Pixl 🥲, but is unrelated to the PR.

Addressed via https://github.com/Automattic/wp-calypso/pull/78924

miksansegundo reviewed (approved):
🎉 LGTM!

When upgrading the plan from the styles modal in the theme preview, I was redirected back to the theme preview with the state preserved so I need to click on `Continue`. 

https://github.com/Automattic/wp-calypso/assets/1881481/efcc19f5-0380-497a-947d-ed8ab4df96bb

I know this is expected and I think we are planning to improve it, maybe showing the celebration page or a more visible upgrade confirmation message because the one on the checkout page is barely visible due to the scroll.

<img width="1094" alt="Screenshot 2566-07-04 at 11 58 04" src="https://github.com/Automattic/wp-calypso/assets/1881481/ff77b617-0f35-42d3-b370-fbd133b75ea2">

arthur791004 committed:
Fix types

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8351117

Thank you @arthur791004 for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8351117).
